keylime/ima/ima_dm.py

```diff
@@ -904,11 +959,24 @@ def mirror_mirror_device_data(children):
 # The parser_fast only works on newer lark versions
 parser_fast = Lark(DM_GRAMMAR, parser="lalr", transformer=transformer)
 
+EventTypes = Union[LoadEvent, ClearEvent, RemoveEvent, ResumeEvent, RenameEvent, UpdateEvent]
 
-def parse(data: str, event: str):
+
+def parse(data: str, event: str) -> EventTypes:
     if version.Version(lark.__version__) >= version.Version("1.0.0"):
         out = parser_fast.parse(event + data)
     else:
         out = transformer.transform(parser.parse(event + data))
-    # TODO: Provide sanity checks for object construction e.g a linear target should always have LinearAttributes
+
+    if (event, type(out)) not in [
+        ("dm_table_load", LoadEvent),
+        ("dm_device_resume", ResumeEvent),
+        ("dm_device_remove", RemoveEvent),
+        ("dm_table_clear", ClearEvent),
+        ("dm_device_rename", RenameEvent),
+        ("dm_target_update", UpdateEvent),
+    ]:
+        raise TypeError(f"{event} was parsed as: {type(event)}")
+
+    assert isinstance(out, (LoadEvent, ResumeEvent, RemoveEvent, ClearEvent, RenameEvent, UpdateEvent))
```

Comment:
The above should have this already covered...

Reply:
Yes, but unfortunately mypy does not find it. I could use cast, but I don't know if this is the better solution.

Reply:
Could you move the type checking code into a function of its own and also apply it to those checks in `validate()`? I think that function could be called from both places? As long as pyright/mypy is happy afterwards...

Reply:
@stefanberger I'll try to find a way to make this more elegant. The issue I don't think there is an elegant way to associate input values with the output type. @aplanas do you have an idea to solve this?

Reply:
No idea!

Only a suggestion: if we create a function `def _from_str_to_even(event: str) -> EventTypes:` could this help?

Reply:
It is probably a good idea to create that function, but I don't think we can get rid of the asserts in the validate function...

Reply:
@aplanas what do you think of the idea of introducing functions of the style `parse_LoadEvent(data: str) -> LoadEvent:`? The advantage is that we have type safety, the disadvantage is that we introduce 6 new functions that basically are doing the same thing.

Reply:
Ok, then we will keep the asserts for now. Maybe in the future we find a better way to solve this.